client/requestUtil.js

```diff
@@ -169,11 +199,216 @@ RequestUtil.prototype.list = function (category, startAt, maxRecords) {
   }
   if (startAt) { options.StartAfter = `${prefix}/${startAt}` }
   return this.withRetry(() => {
-    return s3Helper.listObjects(this.s3, options, !!maxRecords)
+    if (!startAt || startAt === 0 || ((new Date()).getTime() - startAt) >
+        parseInt(s3Helper.SQS_RETENTION, 10) * 1000) {
+      return s3Helper.listObjects(this.s3, options, !!maxRecords)
+    }
+    // We poll from SQS
+    let notificationParams = {
+      QueueUrl: `${this.SQSUrl}`,
+      AttributeNames: [
+        'All'
+      ],
+      MaxNumberOfMessages: SQS_MAX_LIST_MESSAGES_COUNT,
+      MessageAttributeNames: [
+        'All'
+      ],
+      VisibilityTimeout: SQS_MESSAGES_VISIBILITY_TIMEOUT,
+      WaitTimeSeconds: SQS_MESSAGES_LONGPOLL_TIMEOUT
+    }
+
+    return s3Helper.listNotifications(this.SQS, notificationParams, category,
+      `${this.apiVersion}/${encodeURIComponent(this.userId)}/${category}`, platform)
   })
 }
 
 /**
+ * Creates SNS Topic for the current user.
+ * @returns {Promise}
+ */
+RequestUtil.prototype.createAndSubscribeSNS = function () {
+  this.SNSName = `${this.bucket}_sns_${this.userId.replace(/[^A-Za-z0-9]/g, '')}`
+  let params = {
+    Name: `${this.SNSName}`
+  }
+  return new Promise((resolve, reject) => {
+    this.SNS.createTopic(params, (error, data) => {
+      if (error) {
+        console.log('SNS creation failed with error: ' + error)
+        reject(error)
+      } else if (data) {
+        this.SNSArn = data.TopicArn
+        let topicAttributesParams = {
+          AttributeName: 'Policy',
+          TopicArn: `${data.TopicArn}`,
+          AttributeValue: `${this.SNSPolicy(data.TopicArn)}`
+        }
+        this.SNS.setTopicAttributes(topicAttributesParams, (errorAttr, dataAttr) => {
+          if (errorAttr) {
+            console.log('SNS setTopicAttributes failed with error: ' + errorAttr)
+            reject(errorAttr)
+          } else if (dataAttr) {
+            let encodedUser = encodeURIComponent(this.userId)
+            let bucketNotificationConfiguration = {
+              Bucket: `${this.bucket}`,
+              NotificationConfiguration: {
+                TopicConfigurations: [
+                  {
+                    Events: [
+                      's3:ObjectCreated:Post'
```

Comment:
minor: maybe it's not important now, but in the future maybe we need to subscribe to ObjectDeleted as well (happens when user hits Reset Sync which deletes everything from S3)

Reply:
That subscription is: our s3 bucket is allowed to post notification to SNS topic. Probably it is a good idea to delete the corresponding SNS and SQS in that case. Thanks, I'll do it.